Minimal *-deployment dependencies for integration-tests

famod opened the pull request:
Resolves #10859 (once finalized)

See also [discussion on zulip](https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/parallel.20build.20problem.20in.20integration-test-grpc-*) where we have already discussed a few approaches and details.

A violation looks like this:

```
[INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-deployment-deps) @ quarkus-integration-test-cache ---
[WARNING] Rule 0: io.quarkus.enforcer.RequiresMinimalDeploymentDependency failed with message:
The following *-deployment dependencies are missing/configured incorrectly:
    io.quarkus:quarkus-hibernate-orm-panache-deployment:999-SNAPSHOT
    io.quarkus:quarkus-jdbc-h2-deployment:999-SNAPSHOT
    io.quarkus:quarkus-resteasy-jsonb-deployment:999-SNAPSHOT
    io.quarkus:quarkus-smallrye-context-propagation-deployment:999-SNAPSHOT

Each of these dependencies has to be configured as follows to get a consistent build order:
    <dependency>
        <groupId>GROUP-IP</groupId>
        <artifactId>ARTIFACT-ID</artifactId>
        <version>${project.version}</version>
        <type>pom</type>
        <scope>test</scope>
        <exclusions>
            <exclusion>
                <groupId>*</groupId>
                <artifactId>*</artifactId>
            </exclusion>
        </exclusions>
    </dependency>
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
```

This is WIP for now. General things I'd like to discuss:
- [x] Location under `devtools` yay or nay? Seemed to be the best choice. Alternative: `independent-projects`
- [x] General upgrade of enforcer-plugin from M2 to M3 yay or nay? Clear yay from me.
- [x] Should every required dep be printed fully in XML to copy it "blindly" to the pom without much effort?
- [x] Test coverage in the new module via [`maven-invoker-plugin`](https://maven.apache.org/plugins/maven-invoker-plugin/) yay or nay? I tend to yay.
- [x] ~We should consider using this rule also _outside_ of `integration-tests` (via a subsequent PR) to make the _entire_ reactor build order consistent, not just the test part. This would then also allow us to build incrementally, e.g. with [`gitflow-incremental-builder`](https://github.com/vackosar/gitflow-incremental-builder), which would cut down CI (and local) build times massively. For this we need to decide whether those "symbolic" dependencies would be problematic in case they show up in the "userspace".~
  Postponed! Will restart that discussion with another PR.

Obvious tasks left:
- [x] pom.xml cleanup
- [x] pull the rule up to `integration-tests` parent
- [x] add the now required deployment deps to each submodule (the "fun part") and switch level of rule from `WARN` to `ERROR`
- [x] drop "Symbolic" from the pom.xml comments & enrich rule with some javadoc

Any feedback is very much appreciated!

famod committed:
Minimal *-deployment dependencies for integration-tests + enforcer rule

Fixes #10859 by enforcing a consistent build order.

Presence of such minimal dependencies is enforced by the new RequiresMinimalDeploymentDependency rule.

More details: https://github.com/quarkusio/quarkus/pull/11331#issuecomment-680306828

famod commented:
I'll convert this PR to WIP state after the first CI run finishes (with the expected failure).

famod commented:
Btw, #10859 was originally about grpc tests but as it became clear that it is a general problem, I randomly chose the cache module for my first tests.

famod commented on devtools/enforcer-rules/src/main/java/io/quarkus/enforcer/RequiresMinimalDeploymentDependency.java:
```diff
@@ -0,0 +1,146 @@
+        if (!artifact.isResolved()) {
+            // To avoid this "soft exit", explicit resolving would be necessary but that is pretty elaborate in an enforcer rule.
+            // If the build goal is "late" enough, artifacts for the respective scope *will* be resolved automatically.
+            logger.warn("Artifact is not resolved, consider using a later build goal like 'package': " + artifact);
```
Example:
```
$ mvn clean validate -f integration-tests/cache/
[INFO] Scanning for projects...
[INFO]
[INFO] -------------< io.quarkus:quarkus-integration-test-cache >--------------
[INFO] Building Quarkus - Integration Tests - Cache 999-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-clean-plugin:3.1.0:clean (default-clean) @ quarkus-integration-test-cache ---
[INFO]
[INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (enforce) @ quarkus-integration-test-cache ---
[INFO]
[INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-java-version) @ quarkus-integration-test-cache ---
[INFO]
[INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-maven-version) @ quarkus-integration-test-cache ---
[INFO]
[INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-deployment-deps) @ quarkus-integration-test-cache ---
[WARNING] Artifact is not resolved, consider using a later build goal like 'package': io.quarkus:quarkus-junit5:jar:999-SNAPSHOT:test
[WARNING] Artifact is not resolved, consider using a later build goal like 'package': io.quarkus:quarkus-resteasy-jsonb:jar:999-SNAPSHOT:compile
[WARNING] Artifact is not resolved, consider using a later build goal like 'package': io.quarkus:quarkus-hibernate-orm-panache:jar:999-SNAPSHOT:compile
[WARNING] Artifact is not resolved, consider using a later build goal like 'package': io.quarkus:quarkus-cache:jar:999-SNAPSHOT:compile
[WARNING] Artifact is not resolved, consider using a later build goal like 'package': io.quarkus:quarkus-smallrye-context-propagation:jar:999-SNAPSHOT:compile
[WARNING] Artifact is not resolved, consider using a later build goal like 'package': io.quarkus:quarkus-jdbc-h2:jar:999-SNAPSHOT:compile
[WARNING] Artifact is not resolved, consider using a later build goal like 'package': io.quarkus:quarkus-test-h2:jar:999-SNAPSHOT:test
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

The warnings vanish when `test` is used instead of `validate` and the build will fail (because some deployment deps are missing).

This is not ideal, but I'd say it is sufficient for what the rule is supposed to do, which is to enforce the build order primarily in CI.
CI will always use a "late" phase/goal (at least test).

michalszynkiewicz replied:
@famod , do I understand it correctly `test`  vs `validate` makes difference only if you build the whole project and not when you only select `-f integration-tests/cache/`?

famod replied:
It also makes a difference when builing the entire project.
`mvn validate` called from root would then log many warnings, but failing the build just because artifacts have not yet been resolved would be worse, IMO.

To me, this is an edge case, because usually everyone (including CI) is at least using `compile`, `test`, `package` or even `install`.

I could also squash this into a single warning making it less noisy. WDYT?

famod replied:
I squashed it to a single message for less noise, e.g.:
```
[INFO] --- maven-enforcer-plugin:3.0.0-M2:enforce (enforce-deployment-deps) @ quarkus-integration-test-cache ---
[WARNING] Skipping rule RequiresMinimalDeploymentDependency: Artifacts are not resolved, consider using a later build goal like 'package'.
```

famod commented on devtools/enforcer-rules/src/main/java/io/quarkus/enforcer/RequiresMinimalDeploymentDependency.java:
```diff
@@ -0,0 +1,146 @@
+
+    private static final String GROUP_ID_FILTER = "io.quarkus";
+
+    private static final Map<String, Optional<String>> DEPLOYMENT_GAV_CACHE = new ConcurrentHashMap<>();
```
To avoid re-checking for (and eventually re-parsing of) `quarkus-extension.properties` for dependencies that are used multiple times throughout the entire build reactor.

michalszynkiewicz commented on devtools/enforcer-rules/src/main/java/io/quarkus/enforcer/RequiresMinimalDeploymentDependency.java:
```diff
@@ -0,0 +1,146 @@
+        try {
+            project = (MavenProject) helper.evaluate("${project}");
+        } catch (ExpressionEvaluationException e) {
+            throw new IllegalStateException(e);
```
It would be good to have some message for this exception

famod replied:
This should never happen but you are right, adding a message is best practise.

famod replied:
Message added as suggested.

michalszynkiewicz commented on devtools/enforcer-rules/src/main/java/io/quarkus/enforcer/RequiresMinimalDeploymentDependency.java:
```diff
@@ -0,0 +1,146 @@
+        //   or those parents are not relevant
+
+        Map<String, Dependency> directQuarkusDepsByGAV = project.getDependencies().stream()
+                .filter(d -> d.getGroupId().equals(GROUP_ID_FILTER))
```
just a note, I'm wondering if `startsWith` won't be safer here (and in line 55) in case we e.g. wanted to move grpc stuff to `io.quarkus.grpc`

famod replied:
Replaced with `startsWith` as suggested.

michalszynkiewicz commented on integration-tests/cache/pom.xml:
```diff
@@ -67,6 +80,31 @@
                     </execution>
                 </executions>
             </plugin>
+            <plugin>
```
this would eventually go into `integration-tests/pom.xml`, right?

aloubyansky replied:
If we move extension ITs into its own module, e.g. `integration-tests/extensions` we could configure it in one place. FYI @zakkak

famod replied:
> this will eventually go into `integration-tests/pom.xml`, right?

Yes, that's the plan.

> If we move extension ITs into its own module, e.g. `integration-tests/extensions` we could configure it in one place. 

I am probably missing something, but I don't see any real harm letting the rule run for _all_ of `integration-tests`.

famod replied:
Pulled up to `integration-tests` but level has to be switched from WARN to ERROR after every submodule was extended.

famod commented:
@aloubyansky & @michalszynkiewicz Thanks for your feedback so far. Before the next iteration, what's your take on the tasks I've listed? Any objections?

aloubyansky commented:
I don't have objections to your list. Except probably the last one for which I'm still missing the details.

famod commented:
Pushed the second iteration. Resolved most tasks except adjusment of all it-submodules, enforcer update and invoker test coverage.
The rule is now configured in `integration-tests` but for now only with level `WARN`. The cache module now has the complete list of "symbolic" deployment dependencies (we'll see what CI makes of it, but should be fine).

famod commented:
A violation now lloks like this, making it dead simple to copy over the required minimal/symbolic deps to `pom.xml`:
```
[INFO] --- maven-enforcer-plugin:3.0.0-M2:enforce (enforce-deployment-deps) @ quarkus-integration-test-cache ---
[WARNING] Rule 0: io.quarkus.enforcer.RequiresMinimalDeploymentDependency failed with message:
5 *-deployment dependencies are missing/configured incorrectly:
    io.quarkus:quarkus-cache-deployment:999-SNAPSHOT
    io.quarkus:quarkus-hibernate-orm-panache-deployment:999-SNAPSHOT
    io.quarkus:quarkus-jdbc-h2-deployment:999-SNAPSHOT
    io.quarkus:quarkus-resteasy-jsonb-deployment:999-SNAPSHOT
    io.quarkus:quarkus-smallrye-context-propagation-deployment:999-SNAPSHOT

To fix this issue, add the following dependencies to pom.xml:

        <!-- "Symbolic" test dependencies to *-deployment artifacts for consistent build order -->
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-cache-deployment</artifactId>
            <version>${project.version}</version>
            <type>pom</type>
            <scope>test</scope>
            <exclusions>
                <exclusion>
                    <groupId>*</groupId>
                    <artifactId>*</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-hibernate-orm-panache-deployment</artifactId>
            <version>${project.version}</version>
            <type>pom</type>
            <scope>test</scope>
            <exclusions>
                <exclusion>
                    <groupId>*</groupId>
                    <artifactId>*</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-jdbc-h2-deployment</artifactId>
            <version>${project.version}</version>
            <type>pom</type>
            <scope>test</scope>
            <exclusions>
                <exclusion>
                    <groupId>*</groupId>
                    <artifactId>*</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-resteasy-jsonb-deployment</artifactId>
            <version>${project.version}</version>
            <type>pom</type>
            <scope>test</scope>
            <exclusions>
                <exclusion>
                    <groupId>*</groupId>
                    <artifactId>*</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-smallrye-context-propagation-deployment</artifactId>
            <version>${project.version}</version>
            <type>pom</type>
            <scope>test</scope>
            <exclusions>
                <exclusion>
                    <groupId>*</groupId>
                    <artifactId>*</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
```

famod commented:
Pushed third iteration:
- added smoketest via maven-invoker-plugin (does not test corner cases but should suffice)
- fixed/adjusted a few small things in the rule
- updated to enforcer-plugin 3.0.0-M3 (via separate commit just in case it needs to be reverted)

I am now adding all those missing minimal deps (> 100 pom.xml files, yay...) and after that the PR is ready for (final) review.

famod commented:
Pushed final iteration, **all `integration-tests` submodules now have the required minimal/smybolic `*-deployment` dependencies and the new rule is enforcing them**.

I have yet to conduct a final test to verify that the original problem (#10859) is gone but this PR is now **ready for review!**

famod commented:
> I have yet to conduct a final test to verify that the original problem (#10859) is gone

Just tested this: Fix confirmed.

I see a couple of follow-ups once this is merged:
- enforce that no superfluous minimal/symbolic dependencies are defined
- check whether the rule should also run for the rest of the repo
- check whether https://github.com/vackosar/gitflow-incremental-builder can be used to massively reduce average Quarkus CI build times

famod commented on devtools/enforcer-rules/src/main/java/io/quarkus/enforcer/RequiresMinimalDeploymentDependency.java:
```diff
@@ -0,0 +1,183 @@
+
+    private boolean isMinDeploymentDepPresent(String deploymentGAV, String projArtifactKey,
+            Map<String, Dependency> directDepsByGAV) {
+        return deploymentGAV.equals(projArtifactKey) // special case: current project itself is the "required dependency"
```
This is a bit fishy, but I had to introduce this check for at least on it-submodule (forgot the name).
It almost feels like that submodule is violation the "no runtime deps in deployment modules" rule...

Ladicek replied:
I think the rule is opposite, "no deployment deps in runtime modules", and if there's a violation, then I guess we really want to know :-) Though I'm not sure if this indicates violation of that rule.

famod replied:
"no deployment deps in runtime modules" makes sense as well but I meant: #11174

Ladicek replied:
Ah interesting, I didn't know about that one! Thanks.

famod replied:
I just realized that #11174 is about runtime dependencies of _other_ extensions.
It is perfectly valid (and required!) that the deployment module of an extension A has a dependency on the runtime artifact of extension A.
It would not be valid if that deployment module would depend on the runtime artifact of an extension B!

Long story short: This check is not "fishy". I reworded #11174 a little for clarification.

famod commented:
`Quarkus CI / JDK Java 8 - 242 JVM Tests (pull_request)` failed with:
```
2020-08-20T22:29:27.4041316Z [ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 3.629 s <<< FAILURE! - in io.quarkus.vertx.http.shutdown.ShutdownTest
2020-08-20T22:29:27.4042034Z [ERROR] io.quarkus.vertx.http.shutdown.ShutdownTest  Time elapsed: 3.629 s  <<< FAILURE!
2020-08-20T22:29:27.4042473Z org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
2020-08-20T22:29:27.4042872Z 	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
2020-08-20T22:29:27.4043813Z 	at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:40)
2020-08-20T22:29:27.4044918Z 	at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:35)
2020-08-20T22:29:27.4046395Z 	at org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:162)
2020-08-20T22:29:27.4048515Z 	at io.quarkus.vertx.http.shutdown.ShutdownTest$1.run(ShutdownTest.java:56)
```
but I did not change anything in any extension, only in the `integration-tests` (and this test is a unit test).

Ladicek commented on build-parent/pom.xml:
```diff
@@ -27,6 +27,7 @@
         <scala.version>2.12.8</scala.version>
         <scala-plugin.version>4.1.1</scala-plugin.version>
 
+        <version.enforcer.plugin>3.0.0-M3</version.enforcer.plugin>
```
Is this used anywhere?

famod replied:
It is overriding the version from `jboss-parent`, just like the surefire property below.

Ladicek replied:
Ah! Thanks!

Ladicek commented on devtools/enforcer-rules/pom.xml:
```diff
@@ -0,0 +1,68 @@
+        <plugins>
+            <plugin>
+                <artifactId>maven-invoker-plugin</artifactId>
+                <version>3.2.1</version>
```
Care to put this version string into a property? There's a couple of other modules that use Maven Invoker plugin, and perhaps we should centralize its version. Can be an extra PR too, I don't mind.

famod replied:
Good point. I actually had a look first whether other modules are using this plugin and back then I found none.

But there are two "maven-invoker-way" submodules under `integration-tests` that are using this plugin.

Extra PR sounds reasonable since the version in those two modules are already inconsistent.

Ladicek replied:
Agree.

famod replied:
As a reminder I'll leave this open until this PR is merged.

Ladicek commented on devtools/enforcer-rules/src/main/java/io/quarkus/enforcer/RequiresMinimalDeploymentDependency.java:
```diff
@@ -0,0 +1,184 @@
+                    + " *-deployment dependencies are missing/configured incorrectly:\n"
+                    + "    " + missingDeploymentDeps.stream().collect(Collectors.joining("\n    "))
+                    + "\n\nTo fix this issue, add the following dependencies to pom.xml:\n\n"
+                    + "        <!-- \"Symbolic\" test dependencies to *-deployment artifacts for consistent build order -->\n"
```
I'd just remove the `"Symbolic"` part, they are in no way symbolic, they are perfectly real :-) And their purpose is explained by "for consistent build order". We could add something like `(not needed for any other reason)` to the end, but I think that wouldn't be necessary.

famod replied:
Glad you see it this way because I was not entirely happy with "symbolic" vs./+ "minimal" either.

How about replacing "Symbolic" with "Minimal", hinting at the rule class and adding some javadoc to the rule?

famod replied:
> but I assume those changes are just copy&paste of the enforcer output, so should be pretty safe.

They definitely are. 😉

Ladicek replied:
Minimal makes a lot more sense, as the enforcer rule attempts to find a minimal set of `-deployment` dependencies to add. I personally don't think it's necessary, but I'll leave it to you :-)

famod replied:
Done.

Ladicek commented:
The enforcer rule looks good to me, I only added a few minor comments. I didn't really look at the changed POMs, but I assume those changes are just copy&paste of the enforcer output, so should be pretty safe.

famod commented:
> but I assume those changes are just copy&paste of the enforcer output, so should be pretty safe.

They definitely are. 😉

famod commented:
I have another push pending. I'll wait until the general CI problems are resolved.

famod commented:
Test failure in `Windows JDK 11 JVM Tests (pull_request)` seems unrelated (didn't touch `extensions/resteasy/deployment`):
```
2020-08-25T10:23:43.3421123Z [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 5.173 s <<< FAILURE! - in io.quarkus.resteasy.test.IncompletePostTestCase
2020-08-25T10:23:43.3421577Z [ERROR] io.quarkus.resteasy.test.IncompletePostTestCase.testIncompleteWrite  Time elapsed: 3.103 s  <<< ERROR!
2020-08-25T10:23:43.3421860Z java.net.SocketTimeoutException: Read timed out
```

maxandersen commented:
related/duplicate of #7717

maxandersen commented:
> * Location under `devtools` yay or nay? Seemed to be the best choice. Alternative: `independent-projects`

the other places we have things for ide config and tweaks we have in independent-projects ...like ide-config and revapi; thus that seems more right than devtools imo.

gastaldi commented:
+1 for `independent-projects`

maxandersen commented:
from zulip conversation: 

Reasons to do this: 
a) mvn -am and -amd will work for all of quarkus (faster and easier builds) 
b) tools like GIB (https://github.com/vackosar/gitflow-incremental-builder) can reason about our build without special quarkus knowledge

downsides is that we are listing things we technically don't need but it makes existing tooling work to help make things faster.

famod commented:
> from zulip conversation:
> 
> Reasons to do this:
> a) mvn -am and -amd will work for all of quarkus (faster and easier builds)
> b) tools like GIB (https://github.com/vackosar/gitflow-incremental-builder) can reason about our build without special quarkus knowledge
> 
> downsides is that we are listing things we technically don't need but it makes existing tooling work to help make things faster.

Additional takeaways from the [discussion](https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/gitflow-incremental-builder.20(GIB)):
- Building Quarkus in parallel (e.g. `mvn -T4 -Dquickly`) for the first time (no 999-SNAPSHOT artifacts in local repo) does not break anymore in the grpc integration-tests (#10859)
- If minmal deployment dependencies are rolled out to the rest of the repo as well, building in parallel will always use the deployment artifacts from the current reactor instead of possibly using outdated artifacts from local repo
- `<type>pom</type>` is crucial to keep deployment classes away from (test) classpath (the pom file is added to cp, but that is harmless)
- wildcard exclusions are crucial to keep transitive dependencies of the deployment depenency away from the (test) classpath
- classpath can be verified via `mvn org.apache.maven.plugins:maven-dependency-plugin:3.1.2:build-classpath` (or via `mvn -X test` by checking the file that surefire-plugin is using to define the classpath)
- scope `test` is not so important but this way such a minimal dependency is not propagated downstream (which becomes really important once we decide to use this appproach for non-test modules as well)

New TODOs for @famod :
- [ ] move to `independent-projects` 
- [ ] more detailed test in Eclipse (someone else would need to test IntelliJ)
- [ ] enrich commit message

maxandersen commented:
> * move to `independent-projects`

just noticed the enforcer rule tests has dependency on quarkus thus this isn't actually an independent project. Thus I would say `devtools` is the right place; basically a "extension-plugin" sibling.

aloubyansky commented on devtools/enforcer-rules/src/it/smoketest/ext1/deployment/pom.xml:
```diff
@@ -0,0 +1,13 @@
+        <version>1.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>quarkus-enforcer-rules-smoketest-ext1-deployment</artifactId>
```
To be 100% accurate, the deployment module should depend on the runtime one.

famod replied:
Good point, will add it!

aloubyansky commented on devtools/enforcer-rules/src/main/java/io/quarkus/enforcer/RequiresMinimalDeploymentDependency.java:
```diff
@@ -0,0 +1,188 @@
+                .filter(artifact -> "jar".equals(artifact.getType()))
+                .filter(artifact -> artifact.getGroupId().startsWith(GROUP_ID_PREFIX))
+                .filter(artifact -> !artifact.getArtifactId().endsWith(DEPLOYMENT_ARTIFACT_ID_SUFFIX))
+                .collect(Collectors.toMap(this::buildGAVKey, a -> a));
```
One thing to keep in mind for the future is other extension projects (out there in the community) may want to benefit from this too. In which case GROUP_ID_PREFIX should be either configurable to simply removed in favor of a different approach. E.g. these artifacts are supposed to be resolved at this point, aren't they? We could simply check for the presence of the extension descriptor and also filter based on the resolved path (to collect only those that are found in the current workspace).

famod replied:
> E.g. these artifacts are supposed to be resolved at this point, aren't they?

Yes!

> In which case GROUP_ID_PREFIX should be either configurable

Rules can have parameters so this would be pretty simple, yes.

> We could simply check for the presence of the extension descriptor

If I understood you correctly, this would mean we would "open" each jar artifact to check whether it contains a descriptor file. This will come at a price, but maybe it is just a few additional milliseconds in the end.

> also filter based on the resolved path (to collect only those that are found in the current workspace)

Certainly possible but I see problems when building only a subtree of the repo (reactor does not contain all modules in this case).

aloubyansky replied:
I don't think checking for the presence of the descriptor will be a bottle neck. We are actually doing this check everywhere where we are analyzing the classpath.

As to the subtree, you could use Quarkus bootstrap MavenArtifactResolver instead. There we discover the workspace and it'll work consistently for every build phase and won't depend on the module which was targeted by the mvn cli. I've done this trick in other mojos.

famod replied:
> As to the subtree, you could use Quarkus bootstrap MavenArtifactResolver instead

I actually thought about that when I hit that artifact resolve problem but I decided to keep it KISS until we really, really need it.
In the current form, the rule is really "independent".

famod committed:
Update maven-enforcer-plugin from 3.0.0-M2 to -M3

famod commented:
I'm done! Turns out those minimal dependencies do _not_ at all show up in Eclipse under "Maven Dependencies" (per project).
Currently I cannot run tests propertly from Eclipse but that has a different cause.

So, if someone could test with IntelliJ we're done here (except CI)!

gastaldi reviewed (approved):
Looks good to me, good work @famod !

michalszynkiewicz commented:
thanks @famod !
